[ROCm] Unwrap sccache post-build for ROCm compilations.

iotamudelta opened the pull request:
The sccache wrapping strategy causes problems for at-runtime kernel
compilation of MIOpen kernels. We therefore - after the builds of
caffe2/pytorch are complete - unwrap sccache again by moving the clang-9
actual binary back into its original place.

iotamudelta committed:
Unwrap sccache post-build for ROCm compilations.

The sccache wrapping strategy causes problems for at-runtime kernel
compilation of MIOpen kernels. We therefore after the build of
caffe2/pytorch are complete unwrap sccache again by moving the clang-9
actual binary back into its original place.

iotamudelta committed:
Keep using 4 threads for compilation.

bddppq commented on .jenkins/caffe2/build.sh:
```diff
@@ -271,4 +271,11 @@ fi
 pip install --user -b /tmp/pip_install_onnx "file://${ROOT_DIR}/third_party/onnx#egg=onnx"
 
+if [[ $BUILD_ENVIRONMENT == *rocm* ]]; then
+  if [ -e /opt/rocm/hcc/bin/clang-7.0_original ]; then
```
Hmm this will not work once clang version is updated, can we do it in more general way (like globing `clang-*_original`)

bddppq commented on .jenkins/caffe2/build.sh:
```diff
@@ -271,4 +271,11 @@ fi
+if [[ $BUILD_ENVIRONMENT == *rocm* ]]; then
+  if [ -e /opt/rocm/hcc/bin/clang-7.0_original ]; then
+    # runtime compilation of MIOpen kernels manages to crash sccache - hence undo the wrapping
+    mv /opt/rocm/hcc/bin/clang-7.0_original /opt/rocm/hcc/bin/clang-9
```
ditto

ezyang commented:
AFAICT, you install sccache with this code in build.sh

```
  # ROCm CI is using Caffe2 docker images, which needs these wrapper
  # scripts to correctly use sccache.
  if [ -n "${SCCACHE_BUCKET}" ]; then
    mkdir -p ./sccache

    SCCACHE="$(which sccache)"
    if [ -z "${SCCACHE}" ]; then
      echo "Unable to find sccache..."
      exit 1
    fi

    # Setup wrapper scripts
    for compiler in cc c++ gcc g++; do
      (
        echo "#!/bin/sh"
        echo "exec $SCCACHE $(which $compiler) \"\$@\""
      ) > "./sccache/$compiler"
      chmod +x "./sccache/$compiler"
    done

    export CACHE_WRAPPER_DIR="$PWD/sccache"

    # CMake must find these wrapper scripts
    export PATH="$CACHE_WRAPPER_DIR:$PATH"
  fi
```

Can't you just delete the CACHE_WRAPPER_DIR instead of moving binaries around?

bddppq commented:
Oh right removing the wrappers would is better

iotamudelta commented:
I suspect that the existence of the wrapper script causes problems. Removing the caching directory would not solve that, unfortunately.

@bddppq agreed on using a more general approach

ezyang commented:
>  Removing the caching directory would not solve that, unfortunately.

No I don't understand. What I am suggesting is removing the directory that contains the wrapper scripts.

bddppq commented:
@ezyang it’s because miopen uses the hcc (or clang) in the /opt/rocm directory (in fact the wrapper dir created at build time doesn’t have hcc).

iotamudelta committed:
Glob the original name to make it more future-proof.

The wrapping step names the wrapped binary always clang-7.0_original, so
no glob'ing necessary there.

ezyang commented:
Err, is there a reason there is still a hardcoded version number lol

iotamudelta commented:
Yes, there is. See last commit message and https://github.com/pytorch/pytorch/blob/8399197df67f90d7e5cb52d735eb55c803664b35/docker/caffe2/jenkins/common/install_ccache.sh#L34

ezyang commented:
Resigning from review

iotamudelta committed:
Also glob the remaining hardcoded name.

iotamudelta commented:
@bddppq I've also glob'd the original name now. Can you review? Thanks!

iotamudelta committed:
Use sudo.

bddppq reviewed (approved):
I'm accepting this to unblock, but ideally we should make sccache to work with the miopen runtime compilation commands.

iotamudelta commented:
Yeah, fully agreed that this is ugly.
